conda-store-server/conda_store_server/_internal/utils.py

```diff
@@ -187,3 +188,26 @@ def compile_arn_sql_like(
         re.sub(r"\*", "%", match.group(1)),
         re.sub(r"\*", "%", match.group(2)),
     )
+
+
+def retry_on_errors(allowed_retries=1, on_errors=(), logger=None):
```

Comment:
The reason we're seeing this issue is a race condition that happens between

1. The task that inserts packages found from the `repodata.json` that gets downloaded when a channel gets updated (`task_update_conda_channel`), and
2. The preexisting conda packages in the the conda-store prefixes (`action_add_conda_prefix_packages`)

It seems like making this decorator instantly retry the decorated function upon failure is just delaying one of the two competing tasks rather than solving the underlying race condition. The bulk insert above that gets retried seems like an optimization (that's why it's using bulk insert rather than package by package insertion), so is it possible that `action_add_conda_prefix_packages` wouldn't yet be done by the time the retry is made?

Reply:
Yep, I think you're right. We could wait some time before retrying. But maybe it's worth thinking about it a little more to find a better solution :thinking:

Reply:
I was thinking of a few other options:

1. don't run the `action_add_conda_prefix_packages` if the `task_update_conda_channel` task is running. This option isn't so good because if the task fails, the package won't get added and we'll have no way know about the missing data.
2. when updating channels, check and insert each package instead of doing a bulk insert. This would put a lot more load on the database. This will definitely make things less efficient. We are inserting on the order of 10000's packages per architecture. Probably not a good option